source/extensions/filters/listener/tls_inspector/tls_inspector.cc

```diff
@@ -235,6 +244,149 @@ ParseState Filter::parseClientHello(const void* data, size_t len) {
+  CBS cipher_suites;
+  CBS_init(&cipher_suites, ssl_client_hello->cipher_suites, ssl_client_hello->cipher_suites_len);
+
+  bool first = true;
+  while (CBS_len(&cipher_suites) > 0) {
+    uint16_t id;
+    if (!CBS_get_u16(&cipher_suites, &id)) {
+      return;
+    }
+    if (!first) {
+      out << "-";
+    }
+    out << id;
+    first = false;
+  }
+}
+
+void writeExtensions(const SSL_CLIENT_HELLO* ssl_client_hello, std::ostringstream& out) {
+  CBS extensions;
+  CBS_init(&extensions, ssl_client_hello->extensions, ssl_client_hello->extensions_len);
+
+  bool first = true;
+  while (CBS_len(&extensions) > 0) {
+    uint16_t id;
+    CBS extension;
+
+    if (!CBS_get_u16(&extensions, &id) || !CBS_get_u16_length_prefixed(&extensions, &extension)) {
+      return;
+    }
+    if (isNotGrease(id)) {
+      if (!first) {
+        out << "-";
+      }
+      out << id;
+      first = false;
+    }
+  }
+}
+
+void writeEllipticCurves(const SSL_CLIENT_HELLO* ssl_client_hello, std::ostringstream& out) {
+  const uint8_t* ec_data;
+  size_t ec_len;
+  if (SSL_early_callback_ctx_extension_get(ssl_client_hello, TLSEXT_TYPE_supported_groups,
+                                            &ec_data, &ec_len)) {
+    CBS ec;
+    CBS_init(&ec, ec_data, ec_len);
+
+    bool first = true;
+    while (CBS_len(&ec) > 0) {
+      uint16_t id;
+      if (!CBS_get_u16(&ec, &id)) {
+        return;
+      }
+      if (!first) {
+        out << "-";
+      }
+      out << id;
+      first = false;
+    }
+  }
+}
+
+void writeEllipticCurvePointFormats(const SSL_CLIENT_HELLO* ssl_client_hello, std::ostringstream& out) {
+  const uint8_t* ecpf_data;
+  size_t ecpf_len;
+  if (SSL_early_callback_ctx_extension_get(ssl_client_hello, TLSEXT_TYPE_ec_point_formats,
+                                            &ecpf_data, &ecpf_len)) {
+    CBS ecpf;
+    CBS_init(&ecpf, ecpf_data, ecpf_len);
+
+    bool first = true;
+    while (CBS_len(&ecpf) > 0) {
+      uint8_t id;
+      if (!CBS_get_u8(&ecpf, &id)) {
+        return;
+      }
+      if (!first) {
+        out << "-";
+      }
+      out << static_cast<int>(id);
+      first = false;
+    }
+  }
+}
+
+void Filter::connectionFingerprint(const SSL_CLIENT_HELLO* ssl_client_hello) {
+  if (config_->enableConnectionFingerprinting()) {
+    std::ostringstream out;
+    const uint16_t client_version = ssl_client_hello->version;
+    out << client_version << ",";
+    writeCipherSuites(ssl_client_hello, out);
+    out << ",";
+    writeExtensions(ssl_client_hello, out);
+    out << ",";
+    writeEllipticCurves(ssl_client_hello, out);
+    out << ",";
+    writeEllipticCurvePointFormats(ssl_client_hello, out);
+
+    std::string fingerprint = out.str();
+    ENVOY_LOG(trace, "tls:connectionFingerprint(), fingerprint: {}", fingerprint);
+
+    uint8_t buf[MD5_DIGEST_LENGTH];
+    MD5(reinterpret_cast<const uint8_t*>(fingerprint.data()), fingerprint.size(), buf);
```

Comment:
Why md5? As a cryptographic hash, it shouldn't really be used anymore; use sha256 instead. For a non-cryptographic hash, use something in `source/common/common/hash.h` that we already use.

Reply:
MD5 is specified by the JA3 standard, it's not cryptographic but the intent is to use a standard hashing algorithm so the fingerprint can be shared.